connectivity/mqtt/sub_utils.c

```diff
@@ -0,0 +1,125 @@
+    log_error(mqtt_sub_logger_id, "[%s:%d:%s]\n", __func__, __LINE__, "SC_TA_NULL");
+    return SC_MQTT_NULL;
+  }
+
+  cfg->sub_config->recv_message = strdup(message->payload);
+  return SC_OK;
+}
+
+void publish_callback_sub_func(struct mosquitto *mosq, void *obj, int mid, int reason_code,
+                               const mosquitto_property *properties) {
+  mosq_config_t *cfg = (mosq_config_t *)obj;
+  UNUSED(reason_code);
+  UNUSED(properties);
+
+  if ((mid == cfg->general_config->last_mid || cfg->general_config->last_mid == 0)) {
+    mosquitto_disconnect_v5(mosq, 0, cfg->property_config->disconnect_props);
+  }
+}
+
+void message_callback_sub_func(struct mosquitto *mosq, void *obj, const struct mosquitto_message *message,
+                               const mosquitto_property *properties) {
+  mosq_config_t *cfg = (mosq_config_t *)obj;
+
+  if (cfg->sub_config->remove_retained && message->retain) {
+    mosquitto_publish(mosq, &cfg->general_config->last_mid, message->topic, 0, NULL, 1, true);
+  }
+
+  if (cfg->sub_config->retained_only && !message->retain) {
+    if (cfg->general_config->last_mid == 0) {
+      mosquitto_disconnect_v5(mosq, 0, cfg->property_config->disconnect_props);
+    }
+    return;
+  }
+
+  if (message->retain && cfg->sub_config->no_retain) return;
+
+  // Copy the message into `cfg` to allow other functions use the message
+  dump_message(cfg, message);
+
+  // Uncomment the following code would cause: once we received a message, then disconnect the connection.
+  // mosquitto_disconnect_v5(mosq, 0, cfg->property_config->disconnect_props);
+}
+
+void connect_callback_sub_func(struct mosquitto *mosq, void *obj, int result, int flags,
+                               const mosquitto_property *properties) {
+  mosq_config_t *cfg = (mosq_config_t *)obj;
+
+  UNUSED(flags);
+  UNUSED(properties);
+
+  if (!result) {
+    mosquitto_subscribe_multiple(mosq, NULL, cfg->sub_config->topic_count, cfg->sub_config->topics,
+                                 cfg->general_config->qos, 0, cfg->property_config->subscribe_props);
+
+    for (int i = 0; i < cfg->sub_config->unsub_topic_count; i++) {
+      mosquitto_unsubscribe_v5(mosq, NULL, cfg->sub_config->unsub_topics[i], cfg->property_config->unsubscribe_props);
+    }
+  } else {
+    if (result) {
+      if (cfg->general_config->protocol_version == MQTT_PROTOCOL_V5) {
+        log_error(mqtt_sub_logger_id, "[%s:%d]:%s\n", __func__, __LINE__, mosquitto_reason_string(result));
+      } else {
+        log_error(mqtt_sub_logger_id, "[%s:%d]:%s\n", __func__, __LINE__, mosquitto_connack_string(result));
+      }
+    }
+    mosquitto_disconnect_v5(mosq, 0, cfg->property_config->disconnect_props);
+  }
+}
+
+void subscribe_callback_sub_func(struct mosquitto *mosq, void *obj, int mid, int qos_count, const int *granted_qos) {
+  printf("Subscribed (mid: %d): %d", mid, granted_qos[0]);
+  for (int i = 1; i < qos_count; i++) {
+    printf(", %d", granted_qos[i]);
```

Comment:
Are these messages for debugging?

Reply:
This is logging message. It comes from the original repo, `mosquitto`.